core/common/src/Clock.kt

```diff
@@ -40,17 +40,50 @@ public fun Clock.todayIn(timeZone: TimeZone): LocalDate =
  * Returns a [TimeSource] that uses this [Clock] to mark a time instant and to find the amount of time elapsed since that mark.
  */
 @ExperimentalTime
-public fun Clock.asTimeSource(): TimeSource = object : TimeSource {
-    override fun markNow(): TimeMark = InstantTimeMark(now(), this@asTimeSource)
+public fun Clock.asTimeSource(): TimeSource.WithComparableMarks = object : TimeSource.WithComparableMarks {
+    override fun markNow(): ComparableTimeMark = InstantTimeMark(now(), this@asTimeSource)
 }
 
 @ExperimentalTime
-private class InstantTimeMark(private val instant: Instant, private val clock: Clock) : TimeMark {
-    override fun elapsedNow(): Duration = clock.now() - instant
+private class InstantTimeMark(private val instant: Instant, private val clock: Clock) : ComparableTimeMark {
+    override fun elapsedNow(): Duration = saturatingDiff(clock.now(), instant)
 
-    override fun plus(duration: Duration): TimeMark = InstantTimeMark(instant + duration, clock)
+    override fun plus(duration: Duration): ComparableTimeMark = InstantTimeMark(instant.saturatingAdd(duration), clock)
+    override fun minus(duration: Duration): ComparableTimeMark = InstantTimeMark(instant.saturatingAdd(-duration), clock)
 
-    override fun minus(duration: Duration): TimeMark = InstantTimeMark(instant - duration, clock)
+    override fun minus(other: ComparableTimeMark): Duration {
+        if (other !is InstantTimeMark || other.clock != this.clock) {
+            throw IllegalArgumentException("Subtracting or comparing time marks from different time sources is not possible: $this and $other")
+        }
+        return saturatingDiff(this.instant, other.instant)
+    }
+
+    override fun equals(other: Any?): Boolean {
+        return other is InstantTimeMark && this.clock == other.clock && this.instant == other.instant
+    }
+
+    override fun hashCode(): Int = instant.hashCode()
+
+    override fun toString(): String = instant.toString()
+
+    private fun Instant.isSaturated() = this == Instant.MAX || this == Instant.MIN
+    private fun Instant.saturatingAdd(duration: Duration): Instant {
+        if (isSaturated()) {
+            if (duration.isInfinite() && duration.isPositive() != this.isDistantFuture) {
+                throw IllegalArgumentException("Summing infinities of different signs")
```

Comment:
But we already have `Instant.plus(Duration)`. Its behavior *is* slightly different, and I even like `saturatingAdd` more, but shouldn't `saturatingDiff` also throw on NaN if we keep `saturatingAdd` the way it is here?

Reply:
> shouldn't saturatingDiff also throw on NaN

No, it's defined to return zero when calculating difference between two equal infinitely distant time marks.

Reply:
Isn't it internally inconsistent? This would make sense in the `Instant` infinity model of "all events in the far future are just one event," but in the `Mark` infinity model of "infinitely far events are fuzzily distributed along the timeline," this is surprising.

Reply:
It's consistent in time mark infinity model: all the distant events are the same infinitely distant event. It's not consistent with math infinity, but we're ok with that. See the Libraries DM at 2023-04-11.

Reply:
Oh, ok. This model is so unintuitive to me that I didn't even think about it. Well, since the behavior is already established and you implemented it precisely, I guess there's no need to discuss anything further, as my gripes are with the behavior, not the implementation.